File: 2023/19/python/main.py
from __future__ import annotations

from curses import nonl
import fileinput
from collections.abc import Generator
from copy import copy
from fileinput import FileInput
from itertools import takewhile
from math import prod
from typing import Literal, NamedTuple

type Action = str | bool


class Part(dict[str, int]):
    def score(self) -> int:
        return sum(self.values())

    def check(self) -> bool:
        curr = workflows["in"]
        while not isinstance(action := curr.run(self), bool):
            curr = workflows[action]

        return action


class Condition(NamedTuple):
    category: str
    operator: Literal["<"] | Literal[">"]
    limit: int

    def test(self, part: Part) -> bool:
        if (self.operator == ">" and part[self.category] > self.limit) or (
            self.operator == "<" and part[self.category] < self.limit
        ):
            return True

        return False

    def split(self, part_range: PartRange) -> tuple[PartRange, PartRange]:
        accept, reject = copy(part_range), copy(part_range)
        o_range = part_range[self.category]

        match self.operator:
            case "<":
                accept[self.category] = Range(o_range.low, self.limit - 1)
                reject[self.category] = Range(self.limit, o_range.high)
            case ">":
                accept[self.category] = Range(self.limit + 1, o_range.high)
                reject[self.category] = Range(o_range.low, self.limit)

        return accept, reject


class Rule(NamedTuple):
    condition: Condition | None
    action: Action

    def test(self, part) -> Action | None:
        if self.condition is None or self.condition.test(part):
            return self.action


class Range(NamedTuple):
    low: int
    high: int


class PartRange(dict[str, Range]):
    def score(self) -> int:
        return prod(high - low + 1 for low, high in self.values())

    def solve(self) -> list[PartRange]:
        acceptables: list[PartRange] = []

        def solve_workflow(workflow: WorkFlow):
            nonlocal self

            for self, action in workflow.sovle(self):
                print(self, action)

                if action is True:
                    acceptables.append(self)

                if isinstance(action, str):
                    solve_workflow(workflows[action])

        solve_workflow(workflows["in"])
        return acceptables


class WorkFlow(list[Rule]):
    def run(self, part: Part) -> Action:
        for rule in self:
            if (res := rule.test(part)) is not None:
                return res

        raise AssertionError("Non redirecting or terminating workflow")

    def sovle(self, part_range: PartRange) -> Generator[tuple[PartRange, Action]]:
        for condition, action in self:
            if condition is None:
                accept = part_range
            
            else:
                accept, part_range = condition.split(part_range)
            
            yield accept, action


workflows: dict[str, WorkFlow] = {}
parts: list[Part] = []


def parse_input(file: FileInput[str]):
    def parse_workflow(line: str):
        def parse_action(action: str) -> bool | str:
            match action:
                case "A":
                    return True
                case "R":
                    return False
                case _:
                    return action

        def parse_condition(condition: str) -> Condition:
            if any((pos := condition.find(operator := o)) != -1 for o in ("<", ">")):
                return Condition(condition[:pos], operator, int(condition[pos + 1 :]))

            raise AssertionError(f"unknown operator: {operator}")

        def parse_rule(rule: str) -> Rule:
            if (index := rule.find(":")) == -1:
                return Rule(None, parse_action(rule))

            condition, action = rule[:index], rule[index + 1 :]
            return Rule(parse_condition(condition), parse_action(action))

        name = line[: line.index("{")]
        workflows[name] = WorkFlow(
            parse_rule(rule) for rule in line[len(name) + 1 : -2].split(",")
        )

    def parse_part(line: str):
        part = Part()

        for cat_str in line[1:-2].split(","):
            category, amount = cat_str.split("=")
            part[category] = int(amount)

        parts.append(part)

    for line in takewhile(lambda l: len(l) > 1, file):
        parse_workflow(line)

    for line in file:
        parse_part(line)


parse_input(fileinput.input(encoding="utf-8"))

print(f"P1: {sum(p.score() for p in parts if p.check())}")
print(
    f"P2: {sum(pr.score() for pr in PartRange({
    "x": Range(1, 4000),
    'm': Range(1, 4000),
    'a': Range(1, 4000),
    's': Range(1, 4000),
}).solve())}"
)

File: 2023/19/go/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

type rule interface {
	resolveAction() (redirection string, result bool)
}

type action string

func (a action) resolveAction() (redirection string, result bool) {
	switch a {
	case "A":
		return redirection, true
	case "R":
		return redirection, false
	default:
		return string(a), result
	}
}

type condition struct {
	category rune
	operator rune
	limit    int
}

func (c condition) test(p part) bool {
	switch c.operator {
	case '>':
		return p[c.category] > c.limit
	case '<':
		return p[c.category] < c.limit
	default:
		panic("unrecognized operator in condition argument")
	}

}

type conditionRule struct {
	condition
	action
}

type part map[rune]int

func (p part) score() (res int) {
	for _, amount := range p {
		res += amount
	}

	return
}

func (p part) solve(wf workflow) bool {
	for _, rule := range wf {
		switch v := rule.(type) {
		case conditionRule:
			if !v.test(p) {
				continue
			}
		}

		redirection, result := rule.resolveAction()
		if redirection == "" {
			return result
		}

		return p.solve(workflows[redirection])
	}

	panic(fmt.Sprintf("Unresolvable workflow: %v", wf))
}

type workflow []rule

var workflows = map[string]workflow{}
var parts = []part{}

func parseInput(r io.Reader) {
	scanner := bufio.NewScanner(r)

	parseCondition := func(c string) condition {
		limit, _ := strconv.Atoi(c[2:])
		return condition{rune(c[0]), rune(c[1]), limit}
	}

	for scanner.Scan() {
		if len(scanner.Text()) == 0 {
			break
		}

		line := scanner.Text()[:len(scanner.Text())-1]

		fields := strings.Split(line, "{")
		name := fields[0]
		rules := strings.Split(fields[1], ",")

		wf := workflow{}
		for _, r := range rules {
			ruleSplit := strings.Split(r, ":")
			switch len(ruleSplit) {
			case 1:
				wf = append(wf, action(ruleSplit[0]))
			case 2:
				wf = append(wf, conditionRule{parseCondition(ruleSplit[0]), action(ruleSplit[1])})
			}
		}

		workflows[name] = wf
	}

	for scanner.Scan() {
		line := scanner.Text()[1 : len(scanner.Text())-1]

		newPart := part{}
		partSplit := strings.Split(line, ",")
		for _, c := range partSplit {
			amount, _ := strconv.Atoi(c[2:])
			newPart[rune(c[0])] = amount
		}

		parts = append(parts, newPart)
	}
}

func solve() (res int) {
	for _, part := range parts {
		if part.solve(workflows["in"]) {
			res += part.score()
		}
	}

	return
}

func main() {
	parseInput(os.Stdin)
	P1 := solve()
	P2 := 0

	fmt.Printf("P1: %d \n", P1)
	fmt.Printf("P2: %d \n", P2)
}
